ompi/mca/coll/base/coll_base_util.c

```diff
@@ -103,3 +103,165 @@ int ompi_rounddown(int num, int factor)
     num /= factor;
     return num * factor;    /* floor(num / factor) * factor */
 }
+
+static void release_objs_callback(struct ompi_coll_base_nbc_request_t *request) {
+    if (NULL != request->data.objs.objs[0]) {
+        OBJ_RELEASE(request->data.objs.objs[0]);
+    }
+    if (NULL != request->data.objs.objs[1]) {
+        OBJ_RELEASE(request->data.objs.objs[1]);
+    }
+}
+
+static int complete_objs_callback(struct ompi_request_t *req) {
+    struct ompi_coll_base_nbc_request_t *request = (ompi_coll_base_nbc_request_t *)req;
+    int rc = OMPI_SUCCESS;
+    assert (NULL != request);
+    if (NULL != request->cb.req_complete_cb) {
+        rc = request->cb.req_complete_cb(request->req_complete_cb_data);
+    }
+    release_objs_callback(request);
+    return rc;
+}
+
+static int free_objs_callback(struct ompi_request_t **rptr) {
+    struct ompi_coll_base_nbc_request_t *request = *(ompi_coll_base_nbc_request_t **)rptr;
+    int rc = OMPI_SUCCESS;
+    if (NULL != request->cb.req_free) {
+        rc = request->cb.req_free(rptr);
+    }
+    release_objs_callback(request);
+    return rc;
+}
+
+int ompi_coll_base_retain_op( ompi_request_t *req, ompi_op_t *op,
+                              ompi_datatype_t *type) {
+    ompi_coll_base_nbc_request_t *request = (ompi_coll_base_nbc_request_t *)req;
+    bool retain = false;
+    if (!ompi_op_is_intrinsic(op)) {
+        OBJ_RETAIN(op);
+        request->data.op.op = op;
+        retain = true;
+    }
+    if (!ompi_datatype_is_predefined(type)) {
+        OBJ_RETAIN(type);
+        request->data.op.datatype = type;
+        retain = true;
+    }
+    if (OPAL_UNLIKELY(retain)) {
+        if (req->req_persistent) {
```

Comment:
This is a good catch. Please add a detailed comment on why you need to trap either the complete or the free callback.